dockertest/docker_daemon.py

```diff
@@ -112,51 +111,136 @@ def version(self):
+                    docker = 'docker-latest'
+    return docker
 
-    daemon_process = utils.AsyncJob(" ".join(cmd), close_fds=True)
-    return daemon_process
 
+def _systemd_action(action):
+    utils.run("systemctl %s %s.service" % (action, _which_docker()))
+    # FIXME: check status
 
-def output_match(daemon_process,
-                 timeout=120,
-                 regex=r"-job acceptconnections\(\) = OK \(0\)"):
+
+def stop():
+    """ stop the docker daemon """
+    _systemd_action('stop')
+
+
+def start():
+    """ start the docker daemon """
+    _systemd_action('start')
+
+
+def restart():
+    """ restart the docker daemon """
+    _systemd_action('restart')
+
+
+def _preserved_extension():
     """
-    Return True if daemon_process output matches regex within timeout period
+    File extension (including dot) used for our backup files
+    """
+    return '.docker-autotest-preserved'
+
 
-    :param daemon_process: Opaque daemon_process object (not for direct use)
-    :param regex: Regular expression to search for
-    :param timeout: Maximum time to wait before returning False
+def assert_pristine_environment():
     """
-    return wait_for_output(daemon_process.get_stderr,
-                           regex,
-                           timeout=timeout)
+    Barf if there are any leftover .docker-autotest-preserved files
+    in /etc/sysconfig; this would indicate that a previous test
+    failed to clean up properly, and our system is in an
+    undefined state.
+    """
+    for suffix in ['', '-latest']:
+        path = '/etc/sysconfig/docker%s%s' % (suffix, _preserved_extension())
+        if os.path.exists(path):
+            raise RuntimeError("Leftover backup file: %s. System is"
+                               " in undefined state! Please examine that"
+                               " and its original file; if appropriate,"
+                               " mv it back into place." % path)
+
+
+def edit_options_file(remove=None, add=None):
+    """
+    Write a new /etc/sysconfig/docker* file with new OPTIONS string.
+    Preserve the original.
 
+    :param remove: string or list of strings - option(s) to remove from line
+    :param add: string or list of strings - option(s) to add to line
+    """
+    sysconfig_file = '/etc/sysconfig/%s' % _which_docker()
+    sysconfig_bkp = sysconfig_file + _preserved_extension()
+    if os.path.exists(sysconfig_bkp):
+        raise RuntimeError("Backup file already exists: %s" % sysconfig_bkp)
+    sysconfig_tmp = sysconfig_file + '.tmp'
+    with open(sysconfig_file, 'r') as sysconfig_fh_in:
+        with open(sysconfig_tmp, 'w') as sysconfig_fh_out:
+            for line in sysconfig_fh_in:
+                if line.startswith('OPTIONS='):
+                    line = edit_options_string(line, remove=remove, add=add)
+                sysconfig_fh_out.write(line)
+    os.link(sysconfig_file, sysconfig_bkp)
+    os.rename(sysconfig_tmp, sysconfig_file)
```

Comment:
Good we never ever never never ever run more than one Docker Autotest at a time on the same host :grinning: